### PR TITLE
Add hardware configuration option support

### PR DESCRIPTION
**Pull request by mikee47**
This PR aims to simplify use of hardware configurations via option settings. This allows common bits of configuration to be pulled in by name.

See https://github.com/SmingHub/Sming/issues/1712#issuecomment-798885091

### CODE REVIEW

File: Sming/Arch/Esp8266/options.json [hunk 1]
```diff
@@ -0,0 +1,10 @@
+{
+    "vdd": {
+        "description": "Modify PHY configuration to support `system_get_vdd33()` function",
+        "partitions": {
+            "phy_init": {
+                "filename": "$(FLASH_INIT_DATA_VCC)"
+            }
+        }
+    }
+}
```
Comment (mikee47) [inline Sming/Arch/Esp8266/options.json:10]: @slaff Here's the glue for the 'vdd' option, just needs your code to implement `FLASH_INIT_DATA_VCC`

**Comment by slaff** [comment]
@mikee47 Thank you and I like a lot the elegancy and flexibility of the hardware options!

**Comment by slaff** [comment]
@mikee47 I am having problems using the `vdd` option. I have tried on my machine and then katacoda's VM to enable vdd for Basic_Blink using this command 

```
cd $SMING_HOME/../samples/Basic_Blink
make hwconfig HWCONFIG_OPTS=vdd
```

and in both places I get the following error:

```
Basic_Blink: Invoking 'hwconfig' for Esp8266 (debug) architecture
Traceback (most recent call last):
  File "/root/Sming/Sming/Components/Storage/Tools/hwconfig/hwconfig.py", line 114, in <module>
    main()
  File "/root/Sming/Sming/Components/Storage/Tools/hwconfig/hwconfig.py", line 96, in main
    config = Config.from_name(args.input)
  File "/root/Sming/Sming/Components/Storage/Tools/hwconfig/config.py", line 46, in from_name
    config.load(name, [] if options == '' else options.split(','))
  File "/root/Sming/Sming/Components/Storage/Tools/hwconfig/config.py", line 56, in load
    self.parse_dict(data, options)
  File "/root/Sming/Sming/Components/Storage/Tools/hwconfig/config.py", line 78, in parse_dict
    self.parse_options(options)
  File "/root/Sming/Sming/Components/Storage/Tools/hwconfig/config.py", line 71, in parse_options
    self.parse_dict(temp)
  File "/root/Sming/Sming/Components/Storage/Tools/hwconfig/config.py", line 100, in parse_dict
    self.partitions.parse_dict(partitions, self.devices)
  File "/root/Sming/Sming/Components/Storage/Tools/hwconfig/partition.py", line 127, in parse_dict
    part = Entry(devices[0], name)
  File "/root/Sming/Sming/Components/Storage/Tools/hwconfig/storage.py", line 56, in __getitem__
    return super(List, self).__getitem__(item)
IndexError: list index out of range
Traceback (most recent call last):
  File "/root/Sming/Sming/Components/Storage/Tools/hwconfig/hwconfig.py", line 114, in <module>
    main()
  File "/root/Sming/Sming/Components/Storage/Tools/hwconfig/hwconfig.py", line 96, in main
    config = Config.from_name(args.input)
  File "/root/Sming/Sming/Components/Storage/Tools/hwconfig/config.py", line 46, in from_name
    config.load(name, [] if options == '' else options.split(','))
  File "/root/Sming/Sming/Components/Storage/Tools/hwconfig/config.py", line 56, in load
    self.parse_dict(data, options)
  File "/root/Sming/Sming/Components/Storage/Tools/hwconfig/config.py", line 78, in parse_dict
    self.parse_options(options)
  File "/root/Sming/Sming/Components/Storage/Tools/hwconfig/config.py", line 71, in parse_options
    self.parse_dict(temp)
  File "/root/Sming/Sming/Components/Storage/Tools/hwconfig/config.py", line 100, in parse_dict
    self.partitions.parse_dict(partitions, self.devices)
  File "/root/Sming/Sming/Components/Storage/Tools/hwconfig/partition.py", line 127, in parse_dict
    part = Entry(devices[0], name)
  File "/root/Sming/Sming/Components/Storage/Tools/hwconfig/storage.py", line 56, in __getitem__
    return super(List, self).__getitem__(item)
IndexError: list index out of range

Traceback (most recent call last):
  File "/root/Sming/Sming/Components/Storage/Tools/hwconfig/hwconfig.py", line 114, in <module>
    main()
  File "/root/Sming/Sming/Components/Storage/Tools/hwconfig/hwconfig.py", line 96, in main
    config = Config.from_name(args.input)
  File "/root/Sming/Sming/Components/Storage/Tools/hwconfig/config.py", line 46, in from_name
    config.load(name, [] if options == '' else options.split(','))
  File "/root/Sming/Sming/Components/Storage/Tools/hwconfig/config.py", line 56, in load
    self.parse_dict(data, options)
  File "/root/Sming/Sming/Components/Storage/Tools/hwconfig/config.py", line 78, in parse_dict
    self.parse_options(options)
  File "/root/Sming/Sming/Components/Storage/Tools/hwconfig/config.py", line 71, in parse_options
    self.parse_dict(temp)
  File "/root/Sming/Sming/Components/Storage/Tools/hwconfig/config.py", line 100, in parse_dict
    self.partitions.parse_dict(partitions, self.devices)
  File "/root/Sming/Sming/Components/Storage/Tools/hwconfig/partition.py", line 127, in parse_dict
    part = Entry(devices[0], name)
  File "/root/Sming/Sming/Components/Storage/Tools/hwconfig/storage.py", line 56, in __getitem__
    return super(List, self).__getitem__(item)
IndexError: list index out of range
/root/Sming/Sming/Components/Storage/component.mk:73: recipe for target 'hwconfig' failed
make: *** [hwconfig] Error 1
```

My wild guess is that merging of core opts with arch opts is not working. 

Another related issue: The command below also does not work as expected:

```
make HWCONFIG_OPTS=4m,vdd hwconfig 
```

I get the following error:
```
Basic_Blink: Invoking 'hwconfig' for Esp8266 (debug) architecture
Error in partition entry 'phy_init': type/subtype missing
Error in partition entry 'phy_init': type/subtype missing

Error in partition entry 'phy_init': type/subtype missing
/tmp/Sming/Sming/Components/Storage/component.mk:73: recipe for target 'hwconfig' failed
make: *** [hwconfig] Error 2
```

**Comment by mikee47** [comment]
@slaff Fixed by #2267

**Pull request merged**